source/common/upstream/cluster_manager_impl.cc

```diff
@@ -904,41 +904,51 @@ void ClusterManagerImpl::maybePreconnect(
   }
 }
 
-Http::ConnectionPool::Instance*
+absl::optional<HttpPoolData>
 ClusterManagerImpl::ThreadLocalClusterManagerImpl::ClusterEntry::httpConnPool(
     ResourcePriority priority, absl::optional<Http::Protocol> protocol,
     LoadBalancerContext* context) {
   // Select a host and create a connection pool for it if it does not already exist.
   auto ret = connPool(priority, protocol, context, false);
-
-  // Now see if another host should be preconnected.
-  // httpConnPool is called immediately before a call for newStream. newStream doesn't
-  // have the load balancer context needed to make selection decisions so preconnecting must be
-  // performed here in anticipation of the new stream.
-  // TODO(alyssawilk) refactor to have one function call and return a pair, so this invariant is
-  // code-enforced.
-  maybePreconnect(*this, parent_.cluster_manager_state_, [this, &priority, &protocol, &context]() {
-    return connPool(priority, protocol, context, true);
-  });
-  return ret;
+  if (ret == nullptr) {
+    return absl::nullopt;
+  }
+
+  HttpPoolData data;
+  data.create_stream_ = [this, priority, protocol, context,
+                         ret](Http::ResponseDecoder& response_decoder,
+                              Envoy::Http::ConnectionPool::Callbacks& callbacks)
+      -> Envoy::Http::ConnectionPool::Cancellable* {
+    // Now that a new stream is being established, attempt to preconnect.
+    maybePreconnect(*this, parent_.cluster_manager_state_,
+                    [this, &priority, &protocol, &context]() {
+                      return connPool(priority, protocol, context, true);
+                    });
+    return ret->newStream(response_decoder, callbacks);
+  };
```

Comment:
I tend to prefer explicit objects to lambdas so take this advice with a grain of salt :) If HttpPoolData were an object which had a createStream() function on it, it could be constructed with the various members captured by the lambda. I think that would provide a good place to comment on the lifetime of the objects captured. WDYT?